### PR TITLE
ENH: Add infrastructure for sparse in-place interface.

### PR DESCRIPTION
**Pull request by souravsingh**
The PR is adapted from #5775 incorporated with fixes.

### CODE REVIEW

**Comment by souravsingh** [comment]
@perimosocordiae Can you take a look and suggest any improvements needed?

File: benchmarks/benchmarks/sparse.py [hunk 1]
```diff
@@ -207,15 +207,15 @@ def time_matvecs(self, format):
 
     def time_matvecs_inplace_1_0(self, fmt):
         A = self.matrices[fmt]
```
Comment (perimosocordiae) [inline benchmarks/benchmarks/sparse.py]: This line  (and the others identical to it) should be removed.

File: benchmarks/benchmarks/sparse.py [hunk 2]
```diff
@@ -207,15 +207,15 @@ def time_matvecs(self, format):
 
     def time_matvecs_inplace_1_0(self, fmt):
```
Comment (perimosocordiae) [inline benchmarks/benchmarks/sparse.py]: Each of the inplace tests needs to be skipped if `fmt` is DIA or COO. It might be simpler to just make a second benchmark class, `MatvecsInplace`, which only tests CSR/CSC/BSR formats.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

**Comment by souravsingh** [comment]
@perimosocordiae Are there any more fixes needed?

**Comment by perimosocordiae** [comment]
The merge conflict needs to be fixed, but more importantly the tests are failing. They look like real failures, too.

File: scipy/sparse/sparsetools.py [hunk 3]
```diff
@@ -32,22 +32,22 @@ def _deprecated():
 # of rows.
 
 def csr_matvec(N, M, intptr, indices, data, x, y):
-    _sparsetools.csr_matvec(N, M, indptr, indices, data, 1, x, 0, y)
+    _sparsetools.csr_matvec(N, M, indptr, indices, data, 1, x, 1, y)
```
Comment (ewmoore) [inline scipy/sparse/sparsetools.py]: Can you please explain why this change is correct?  Before these changes the contents of the `y` vector didn't matter.  Making these be `1` means that the values stored in the y vector before this call are now relevant.  This defeats the purpose of wrapping these to try and provide the old interface.

Comment (souravsingh) [inline scipy/sparse/sparsetools.py]: I remember a comment was made on the PR #5775 regarding this. I can change it if needed.

**Comment by souravsingh** [comment]
@perimosocordiae Merge conficts are now resolved.

**Comment by perimosocordiae** [comment]
Looks like you have build failures. Let me know if you need help debugging them.

**Comment by souravsingh** [comment]
@perimosocordiae I would like some help with the debugging of the errors.

**Comment by souravsingh** [comment]
Ping @ewmoore @perimosocordiae

**Comment by perimosocordiae** [comment]
Here's the error I'm seeing in the TravisCI logs:

```
error: no matching function for call to
csr_matvec(const npy_int32&, const npy_int32&, const npy_int32*, const npy_int32*, const npy_bool_wrapper*, const npy_bool_wrapper*, npy_bool_wrapper*)
note: candidate is:
template<class I, class T> void csr_matvec(I, I, const I*, const I*, const T*, T, const T*, T, T*)
```

It looks like you added some of the new function signatures to `generate_sparsetools.py`, but not all of them.
